gr-utils/python/blocktool/cli/base.py

```diff
@@ -0,0 +1,126 @@
+
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import os
+import sys
+from importlib import import_module
+from logging import StreamHandler
+from pkg_resources import iter_entry_points
+
+import click
+from click import ClickException
+from click_plugins import with_plugins
+
+
+class BlockToolException(ClickException):
+    """ Exception class for enhanced CLI interface """
+
+    def show(self, file=None):
+        """ displays the colored message """
+        click.secho('BlockToolException: {}'.format(
+            self.format_message()), fg='red')
+
+
+class ClickHandler(StreamHandler):
+    """
+    This is a derived class of implemented logging class
+    StreamHandler which overrides some of its functional
+    definitions to add colors to the stream output
+    """
+
+    def emit(self, record):
+        """ Writes message to the stream """
+        colormap = {
+            'DEBUG': ('white', 'black'),
+            'INFO': ('blue', None),
+            'WARNING': ('yellow', None),
+            'ERROR': ('red', None),
+            'CRITICAL': ('white', 'red'),
+        }
+        try:
+            msg = self.format(record)
+            colors = colormap.get(record.levelname, (None, None))
+            fgcolor = colors[0]
+            bgcolor = colors[1]
+            click.secho(msg, fg=fgcolor, bg=bgcolor)
+            self.flush()
+        except Exception:
+            self.handleError(record)
+
+
+class CommandCLI(click.Group):
+    """
+    This is a derived class of the implemented click class
+    which overrides some of the functional definitions for external
+    plug-in support
+    """
+    cmd_folder = os.path.abspath(os.path.dirname(__file__))
+
+    def list_commands(self, ctx):
+        """
+        Lists all the commands available in the blocktool directory
+        """
+        cmds = []
+        for filename in os.listdir(self.cmd_folder):
+            if filename.endswith('.py'):
+                cmds.append(filename[:-3])
+        cmds.remove('base')
+        cmds.remove('__init__')
+        cmds += self.commands
+        return sorted(cmds)
+
+    def get_command(self, ctx, cmd_name):
+        """
+        Returns a command object if it exists. The existing in-tree BlockTool
+        command is the priority over the same external plug-in command.
+        """
+        try:
+            if sys.version_info[0] == 2:
+                cmd_name = cmd_name.encode('ascii', 'replace')
+            mod = import_module('gnuradio.blocktool.cli.' + cmd_name)
+        except ImportError:
+            return self.commands.get(cmd_name)
+        return mod.cli
+
+
+@with_plugins(iter_entry_points('gnuradio.blocktool.cli.plugins'))
+@click.command(
+    cls=CommandCLI,
+    epilog='A tool to parse block header files. ' +
+    'Call it without options to run specified command interactively'
+)
+def cli():
+    """Block header parsing tool."""
+    pass
```

Comment:
IMO, plugin support and subcommands are too much. Especially if there is only one...
I propose to strip all the boilerplate stuff and trim this down to one module `cli.py` 

click is fine, but I don't see a need for plugins...